### PR TITLE
Add PatternNet Dataset

### LINKED ISSUE #110: Add PatternNet dataset
[PatternNet](https://sites.google.com/view/zhouwx/dataset) dataset from ["PatternNet: A Benchmark Dataset for Performance Evaluation of Remote Sensing Image Retrieval", Zhou et al. (2018)](https://arxiv.org/abs/1706.03424).

This dataset is a fairly basic remote sensing image/scene classification dataset, however it has some interesting classes which are not found in many of the other scene classification datasets (e.g. oil well, transformer station, christmas tree farm (?), etc.)

patternnet2

### PR DESCRIPTION
**Pull request by isaaccorley**
- Adds `torchgeo.datasets.PatternNet`
- Updated docs
- Added `tests/datasets/test_patternnet.py` (currently passing)
- Added sample data for tests `tests/data/patternnet/PatternNet.zip`

TODO:
- Still can't seem to auto download via google drive but it works with the sample data and the dataset locally.

Questions:
- I'm subclassing from `torchgeo.datasets.VisionDataset` AND `torchvision.datasets.ImageFolder` to reduce boilerplate globbing the files, loading the images. Is this fine or do you prefer we implement the boilerplate manually?

Once this PR is figured out, I'll be adding other classification datasets using the same setup, so I'm interested in getting this one right first.

Closes #110

### CODE REVIEW

**Comment by adamjstewart** [comment]
> Still can't seem to auto download via google drive but it works with the sample data and the dataset locally.

That's fine, we'll wait until torchvision cuts a new release.

> I'm subclassing from torchgeo.datasets.VisionDataset AND torchvision.datasets.ImageFolder to reduce boilerplate globbing the files, loading the images. Is this fine or do you prefer we implement the boilerplate manually?

I'm fine with that. I don't think we'll be able to remove our dependency on torchvision any time soon. If we ever convert this to a GeoDataset we may need to change this, but we can do that at a later date.

File: torchgeo/datasets/patternnet.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    31. solar_panel
+    32. sparse_residential
+    33. storage_tank
+    34. swimming_pool
+    35. tennis_court
+    36. transformer_station
+    37. wastewater_treatment_plant
+
+    If you use this dataset in your research, please cite the following paper:
+    * https://www.sciencedirect.com/science/article/abs/pii/S0924271618300042
```
Comment (adamjstewart) [inline torchgeo/datasets/patternnet.py:73]:
```suggestion
    * https://doi.org/10.1016/j.isprsjprs.2018.01.004
```
I prefer to use the DOI when possible, it's shorter

File: torchgeo/datasets/patternnet.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+        Returns:
+            data and label at that index
+        """
+        image, label = ImageFolder.__getitem__(self, index)
+
+        if self.transforms is not None:
+            image = self.transforms(image)
+
+        sample = {"image": image, "label": label}
```
Comment (adamjstewart) [inline torchgeo/datasets/patternnet.py:136]: `self.transforms` is designed to take in a sample dict as input, not a single image.

File: torchgeo/datasets/patternnet.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    37. wastewater_treatment_plant
+
+    If you use this dataset in your research, please cite the following paper:
+    * https://www.sciencedirect.com/science/article/abs/pii/S0924271618300042
+    """
+
+    url = "https://drive.google.com/file/d/127lxXYqzO6Bd0yZhvEbgIfz95HaEnr9K"
+    md5 = "96d54b3224c5350a98d55d5a7e6984ad"
+    filename = "PatternNet.zip"
+    files = ["images"]
```
Comment (adamjstewart) [inline torchgeo/datasets/patternnet.py:79]:
```suggestion
    directories = ["images"]
```

**Review by adamjstewart — COMMENTED** [review]
Some of the ImageFolder superclass stuff, particularly the transform stuff, seems a bit hacky. In other datasets, we convert from PIL to torch ourselves instead of relying on PILtoTensor. We should standardize on something.

File: torchgeo/datasets/patternnet.py (hunk 2 above)
Comment (isaaccorley) [reply torchgeo/datasets/patternnet.py:136]: Do we want it to act this way for image classification datasets? I understand why it would be this way for detection/segmentation. This seems like it would force the user to roll their own transforms instead of just using transforms = T.Compose([ToTensor(), T.Normalize()])

File: torchgeo/datasets/patternnet.py (hunk 3 above)
Comment (isaaccorley) [reply torchgeo/datasets/patternnet.py:79]: There is only one directory so I need to change it to directory = "images"

File: torchgeo/datasets/patternnet.py (hunk 2 above)
Comment (adamjstewart) [reply torchgeo/datasets/patternnet.py:136]: We want the transforms provided with TorchGeo to work for all datasets, regardless of what type of task we are trying to solve. See https://discuss.pytorch.org/t/state-of-the-art-for-torchvision-datasets-transforms-models-design/123625 for some discussion on this. At the moment, our transform support is pretty poor, but I think my next task will be wrapping the transforms that come with Kornia so that users can use them directly. I'll also be adding geospatial-specific transforms.

Comment (isaaccorley) [reply torchgeo/datasets/patternnet.py:136]: Ok that's fine, I can change it back to use a dict input.

Kornia is awesome. I've used it a ton. But any of their transforms should probably be used after a batch is loaded, no? So that you can use the GPU to perform the normalization and augmentations.

**Comment by isaaccorley** [comment]
Agreed on the hackiness of subclassing ImageFolder. Probably needs a more in depth discussion.

File: torchgeo/datasets/patternnet.py (hunk 2 above)
Comment (adamjstewart) [reply torchgeo/datasets/patternnet.py:136]: That's another thing we need to think about. Right now, ZipDataset is designed in such a way that you pass the transforms to each individual GeoDataset separately, which means that things like RandomHorizontalFlip won't work correctly for Landsat + CDL. One possible solution is to compute the transform after the batch is loaded, but then you can't benefit from data loader parallelization. I haven't yet used Kornia myself so I still need to look into this.

Comment (isaaccorley) [reply torchgeo/datasets/patternnet.py:136]: In my experience it's actually more efficient to take advantage of the multiprocessing in the dataloader only for loading images, and then utilize kornia on the GPU to perform all normalizations & augmentations on a batched tensor. Particularly in the self-supervised learning area where it's common to chain multiple augmentations that significantly slow down CPU data loading (to grayscale + random flips + random resized crop + color jitter). 

The other nice features of kornia which I have been using recently are:
- [kornia.augmentation.VideoSequential](https://kornia.readthedocs.io/en/latest/augmentation.container.html#kornia.augmentation.container.VideoSequential): repeat or vary augmentation per image in time series/video
- [kornia.augmentation.PatchSequential](https://kornia.readthedocs.io/en/latest/augmentation.container.html#kornia.augmentation.container.PatchSequential): repeat or vary augmentation per patch in image
- [kornia.contrib.ExtractTensorPatches](https://kornia.readthedocs.io/en/latest/contrib.html#kornia.contrib.ExtractTensorPatches): extract patches from each image in batch, useful for vision datasets where the images are very large (e.g. 2000x2000) but you want to train on patches (e.g. 256x256)

If using a PyTorch Lightning DataModule for training you can add the augmentations as an input arg and easily use them in the [LightningDataModule.on_after_batch_transfer](https://pytorch-lightning.readthedocs.io/en/latest/extensions/datamodules.html#on-after-batch-transfer) method.

File: torchgeo/datasets/patternnet.py (hunk 3 above)
Comment (isaaccorley) [reply torchgeo/datasets/patternnet.py]: Fixed

File: torchgeo/datasets/patternnet.py (hunk 2 above)
Comment (isaaccorley) [reply torchgeo/datasets/patternnet.py]: Updated to pass sample dict of image and label to transforms. I read the link you posted on the torch forums and I agree this is the way we should standardize.

File: torchgeo/datasets/patternnet.py (hunk 1 above)
Comment (isaaccorley) [reply torchgeo/datasets/patternnet.py]: Updated

**Review by adamjstewart — APPROVED** [review]
